[AN-Issue-1409] Handled automation-registry state update on new epoch

aregng opened the pull request:
- split automation registry external API and internal state to avoid circular dependencies
 - enabled automation-registry-state update on new epoch from block module.

Commit:
Handled automation-registry state update on new epoch

- split automation registry external API and internal state to avoid circular dependencies
- enabled automation-registry-state update on new epoch from block module.

Commit:
Addressed review comments

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -111,74 +71,33 @@ module supra_framework::automation_registry {
-        id: u64
-    }
 
     // todo : this function should call during initialzation, but since we already done genesis in that case who can access the function
```
This can still be called by governance, after framework upgrade by governance.

Commit:
Added couple of unit-tests for automation-registry-state

nizam-supraoracles commented on aptos-move/framework/supra-framework/sources/automation_registry_state.move:
```diff
@@ -221,4 +230,116 @@ module supra_framework::automation_registry_state {
+    }
+
+    #[test(framework = @supra_framework)]
+    #[expected_failure(abort_code = 2)]
```
Always include **location** in failure test to avoid warnings, so it's looks like this `#[expected_failure(abort_code = 2, location = Self)]`.

aregng replied:
Done

nizam-supraoracles commented on aptos-move/framework/supra-framework/sources/automation_registry_state.move:
```diff
@@ -0,0 +1,345 @@
+
+        timestamp::update_global_time_for_test_secs(50);
+        on_new_epoch(30 * MICROSECS_CONVERSION_FACTOR);
+        // Committed gas for the next epoch only for 2 tasks 0 and 2, the task 3 will not be active during upcoming epoch
```
I noticed that you are using the current module's register function for the unit test case.

However, the register function in the `automation_registry` module includes some additional assertion checks that are not being considered in this test.
I believe we should move those assertion checks to the `automation_registry_state` module to ensure consistency.

aregng replied:
agree, updated and added more unit tests.

Commit:
Updated automation-registry to accept registration transaction hash as argument

- Moved basic parameters checks of registration to automation_registry_state
- Added and updated unit tests of autmation_registry_state

nizam-supraoracles commented on aptos-move/framework/supra-framework/sources/transaction_context.move:
```diff
@@ -132,14 +132,6 @@ module supra_framework::transaction_context {
 
-    /// Returns the original transaction hash calculated on the raw-bytes.
-    /// This function aborts if called outside of the transaction prologue, execution, or epilogue phases.
-    public fun txn_app_hash(): vector<u8> {
-        assert!(features::transaction_context_extension_enabled(), error::invalid_state(ETRANSACTION_CONTEXT_EXTENSION_NOT_ENABLED));
-        txn_app_hash_internal()
-    }
-    native fun txn_app_hash_internal(): vector<u8>;
```
Okay, if we remove it from here, I want to know where we will get the transaction_hash from. I am assuming it is handled somewhere in the Rust layer within the CLI.

aregng replied:
It is passed as argument to `automation_registry::register` function from Native/Rust layer when automation-registration transaction is processed, and `register` function parameters are prepared
- https://github.com/Entropy-Foundation/aptos-core/blob/7af8dae5ce9e35ae0eef7ab47733c0edd24ac9c3/aptos-move/framework/supra-framework/sources/automation_registry.move#L162
- https://github.com/Entropy-Foundation/aptos-core/blob/7af8dae5ce9e35ae0eef7ab47733c0edd24ac9c3/aptos-move/aptos-vm/src/aptos_vm.rs#L977
- https://github.com/Entropy-Foundation/aptos-core/blob/7af8dae5ce9e35ae0eef7ab47733c0edd24ac9c3/types/src/transaction/automation.rs#L39

nizam-supraoracles commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -4,18 +4,18 @@
     use supra_framework::timestamp;
     use supra_framework::transaction_context;
+    #[test_only]
+    use std::vector;
```
this imports are unused now

nizam-supraoracles reviewed (approved):
I have just one question regarding the removal of txn_app_hash; everything else looks good to me.

aregng committed:
Removed redundant imports.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry_state.move:
```diff
@@ -0,0 +1,452 @@
+
+    /// It tracks entries both pending and completed, organized by unique indices.
+    struct AutomationRegistryState has key, store {
+        /// A collection of automation task entries that are active state.
+        tasks: EnumerableMap<u64, AutomationTaskMetaData>,
+        current_index: u64,
+        gas_committed_for_next_epoch: u64,
+        automation_gas_limit: u64,
+    }
```
Will all governance configurable parameter also go here? Or are we going to have a separate `config/automation_config.move` similar to `staking_config`? For example, upper limit on duration, base fee, `congestion_threshold`, `congestion_multiplier` all should be governance configurable parameters.

aregng replied:
at this moment parameters has been split into 2 contexts `automation_registry` and `automation_registry_state` based on their place of utilization. but I think as long as we are going to have more and more configurable parameters it would be better to have dedicated structure for it. But I would suggest to do it in a separate PR e.g in scope of the PR wehn automation-fee related configurations  will be introduced.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry_state.move:
```diff
@@ -0,0 +1,452 @@
+                task.is_active = true;
+            }
+        });
+        assert!(expired_task_gas <= state.gas_committed_for_next_epoch, EINVALID_COMMITTED_GAS_CALCULATION);
```
Why does this assertion have to hold? `expired_task_gas` also includes gas for tasks that are expired in previous epoch `task.expiry_time <= current_time`

aregng replied:
After some changes timestamp value units were not taken into account and the committed gas for the next epoch was incorrectly calculated so I have added this assert, but I will try to think of a way to make sure that tests are covering this assert and will remove it.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry_state.move:
```diff
@@ -0,0 +1,452 @@
+        let registry_data = borrow_global_mut<AutomationRegistryState>(@supra_framework);
+        let registration_time = timestamp::now_seconds();
+
+        assert!(expiry_time > registration_time, EINVALID_EXPIRY_TIME);
```
`expiry_time` should at least be in the next epoch

aregng replied:
it has been checked in scope of `automation_registry::register`

sjoshisupra replied:
In that case, we need not check it here.

aregng replied:
Agreed offline that no changes are required here.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry_state.move:
```diff
@@ -0,0 +1,452 @@
+        let registration_time = timestamp::now_seconds();
+
+        assert!(expiry_time > registration_time, EINVALID_EXPIRY_TIME);
+        assert!(gas_price_cap > 0, EINVALID_GAS_PRICE);
```
`gas_price_cap` should be above chain minimum

aregng replied:
I will double check, but chain minimum is not accessible in move layer, even for user-transactions gas-checks are done in Rust/Native layer. 
We can add an extra checks in rust layer before actually calling register at least. I forgot to double check this with you.

sjoshisupra replied:
Perhaps we can pass this parameters to `register` method?

aregng replied:
Also an option, the benefit will be that all checks will be in the same place. But from design point of view, that information has nothing to do with actual registration and only passed to do the checks which could have been done in rust-layer as well.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry_state.move:
```diff
@@ -0,0 +1,452 @@
+
+        assert!(expiry_time > registration_time, EINVALID_EXPIRY_TIME);
+        assert!(gas_price_cap > 0, EINVALID_GAS_PRICE);
+        assert!(max_gas_amount > 0, EINVALID_MAX_GAS_AMOUNT);
```
`max_gas_amount` should be below chain max gas limit

aregng replied:
The same here.

sjoshisupra replied:
Same suggestion, shall we pass this as parameter to `register`? That way these checks would be charged gas to the user.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry_state.move:
```diff
@@ -0,0 +1,452 @@
+        assert!(gas_price_cap > 0, EINVALID_GAS_PRICE);
+        assert!(max_gas_amount > 0, EINVALID_MAX_GAS_AMOUNT);
+        assert!(vector::length(&tx_hash) == TXN_HASH_LENGTH, EINVALID_TXN_HASH);
+        assert!(!vector::is_empty(&payload_tx), EINVALID_PAYLOAD);
```
This assertion may not be required since this would never be called from outside but by the VM, if payload is empty it would fail during payload validation.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry_state.move:
```diff
@@ -0,0 +1,452 @@
+        event::emit(automation_task_metadata);
+    }
+
+    /// Remove Automatioon task entry.
```
```suggestion
    /// Remove Automation task entry.
```

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry_state.move:
```diff
@@ -0,0 +1,452 @@
+    }
+
+    /// Remove Automatioon task entry.
+    public (friend) fun remove_task(owner: &signer, id: u64): AutomationTaskMetaData acquires AutomationRegistryState {
```
`remove_task` can be `public entry fun` since it can be/should be called by the owner

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry_state.move:
```diff
@@ -0,0 +1,452 @@
+    }
+
+    /// Update Automation gas limit
+    public (friend) fun update_automation_gas_limit(
```
It seems that this update has to be followed by `reconfiguration` since otherwise invariant on `automation_gas_limit` as opposed to `gas_committed_for_next_epoch` may be violated. Alternatively,
such an update must fail if `gas_committed_for_next_epoch` goes below `automation_gas_limit`

aregng replied:
This is another good point, we need to have defined policy when these type of updates should actually take place ,  in the same epoch or next epoch, and what should be consequence of it. I will create a ticket for it to have it covered.

sjoshisupra replied:
I believe that perhaps `supra_framework` should have the ability to pause registration of new tasks, that we know that eventually `committed_gas` would be low enough if one wishes to reduce `automation_gas_limit`

aregng replied:
https://github.com/Entropy-Foundation/smr-moonshot/issues/1424

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry_state.move:
```diff
@@ -0,0 +1,452 @@
+        let active_task_ids = vector[];
+        let ids = enumerable_map::get_map_list(&state.tasks);
+
+        vector::for_each(ids, |id| {
+            let task = enumerable_map::get_value(&state.tasks, id);
+            if (task.is_active) {
+                vector::push_back(&mut active_task_ids, id);
+            };
+        });
```
Use `vector::filter`

aregng replied:
I think it is not applicable here as the actual data is got from enumerable-map

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -111,74 +66,33 @@ module supra_framework::automation_registry {
-    }
 
     /// Withdraw accumulated automation task fees from the resource account - access by admin
     entry fun withdraw_automation_task_fees(
```
Should we have a multisig admin instead of demanding that only `supra_framework` can withdraw?

aregng replied:
isn't `supra_framework` already multi-sig admin ? I mean in'st anything that updated by governance utilizing `supra_framework` account. 
I have checked for example `stake_config` updates , they are all done using `supra_framework`. But I can see that non of the functions are entry functions, I assume they can be only utilized in scope of the scripts  which are executed as multisig-transactions. Is this assumption correct. If yes then maybe we should also make the configuration update interfaces as public-functions which can be called only in scope of multi-sig transactions ?

sjoshisupra replied:
Yes, any governance action can only be taken through a script, so the interface would be public functions that checks that signer is indeed `supra_framework`

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -245,69 +155,43 @@ module supra_framework::automation_registry {
-
-        let expiry_time_duration = expiry_time - current_time;
-        assert!(expiry_time_duration < registry_data.duration_upper_limit, EEXPIRY_TIME_UPPER);
+        let task_duration = expiry_time - current_time;
```
subtraction is not safe

aregng replied:
True, I have moved check to registry_state and did not take this into account.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -317,7 +201,7 @@ module supra_framework::automation_registry {
     ) acquires AutomationRegistry {
         let current_time = timestamp::now_seconds();
-        let expiry_time_duration = automation_task_metadata.expiry_time - current_time;
+        let expiry_time_duration = automation_registry_state::task_expiry_time(&automation_task_metadata) - current_time;
```
//TODO: refund calculation based on prevalent congestion fee

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -366,9 +241,8 @@ module supra_framework::automation_registry {
 
     #[view]
     /// Ge gas committed for next epoch
```
```suggestion
    /// Get gas committed for next epoch
```

sjoshisupra commented on aptos-move/framework/supra-framework/sources/block.move:
```diff
@@ -235,6 +236,7 @@ module supra_framework::block {
         randomness::on_new_block(&vm, epoch, round, option::none());
         if (timestamp - reconfiguration::last_reconfiguration_time() >= epoch_interval) {
             reconfiguration::reconfigure();
+            automation_registry_state::on_new_epoch(epoch_interval);
```
This should be called from `reconfiguration`

aregng replied:
reconfiguration does not have access to epoch-interval, either `reconfiguration::reconfigure()` should be updated to accept epoch_interval as parameter or we should leave it as it is.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/block.move:
```diff
@@ -264,6 +266,7 @@ module supra_framework::block {
 
         if (timestamp - reconfiguration::last_reconfiguration_time() >= epoch_interval) {
             reconfiguration_with_dkg::try_start();
+            automation_registry_state::on_new_epoch(epoch_interval);
```
should be called from `reconfiguration`

Commit:
Addressed review comments

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -245,69 +157,46 @@ module supra_framework::automation_registry {
         let registry_data = borrow_global_mut<AutomationRegistry>(@supra_framework);
 
-        // todo : well formedness check of payload_tx
+        //Well formedness check of payload_tx is done in native layer beforehand.
```
```suggestion
        //Well-formedness check of payload_tx is done in native layer beforehand.
```

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry_state.move:
```diff
@@ -0,0 +1,623 @@
+        id: u64
+    }
+
+    public fun task_expiry_time(task: &AutomationTaskMetaData): u64 {
```
```suggestion
   #[view]
     public fun task_expiry_time(task: &AutomationTaskMetaData): u64 {
```

aregng replied:
this is internal function and not a view function exposed to user.

sjoshisupra replied:
I believe that a `[view]` function can only call other `[view]` functions. Therefore, even if we want to call this from other module, this should be declared as `[view]` function. If we do not want to expose this to public, perhaps this can be `public(friend)`

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry_state.move:
```diff
@@ -0,0 +1,623 @@
+            let task = enumerable_map::get_value_mut(&mut state.tasks, id);
+
+            // Tasks that are active during next epoch and are not cancled
+            if (task.state != CANCELLED && task.expiry_time > (current_time + epoch_interval_secs) ) {
+                gas_committed_for_next_epoch = gas_committed_for_next_epoch + task.max_gas_amount;
+            };
+
+            // Drop or activate task
+            if (task.expiry_time <= current_time || task.state == CANCELLED) {
+                enumerable_map::remove_value(&mut state.tasks, id);
+            } else {
+                task.state = ACTIVE;
+            }
+        });
```
```suggestion
            if (task.state != CANCELLED && task.expiry_time > (current_time + epoch_interval_secs) ) {
                gas_committed_for_next_epoch = gas_committed_for_next_epoch + task.max_gas_amount;
                 task.state = ACTIVE;
            }

            else { 
                enumerable_map::remove_value(&mut state.tasks, id);
            }; 
               
            
        });
```

aregng replied:
This function is called when a chain enters a new epoch, and current time will show the start of the new epoch.
with the 1st `if` statement we are collecting committed-gas-amount for the next epoch already.
With the second `if` statement we are clearing up already expired or canceled tasks for this current epoch.
So with the suggested change above we will remove also the tasks that are valid for the the current epoch but expired for the next upcoming epoch.

sjoshisupra replied:
Ah, I missed addition of `epoch_interval_secs`. Fair enough, we can keep it as it is.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry_state.move:
```diff
@@ -0,0 +1,623 @@
+
+        let automation_task_metadata = enumerable_map::get_value(&state.tasks, id);
+        assert!(automation_task_metadata.owner == signer::address_of(owner), EUNAUTHORIZED_TASK_OWNER);
+        assert!(automation_task_metadata.state != CANCELLED, EINVALID_CANCELLATION);
```
```suggestion
        assert!(automation_task_metadata.state != CANCELLED, EALREADY_CANCELLED);
```

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry_state.move:
```diff
@@ -0,0 +1,623 @@
+        // Adjust the gas committed for the next epoch by subtracting the gas amount of the cancelled task
+        state.gas_committed_for_next_epoch = state.gas_committed_for_next_epoch - automation_task_metadata.max_gas_amount;
+
+        event::emit(CanclledAutomationTask { id: automation_task_metadata.id });
```
```suggestion
        event::emit(CancelledAutomationTask { id: automation_task_metadata.id });
```
Upon cancellation some event should also emit information about amount refunded

aregng replied:
I assume when coin transfer happens, there will be deposit event, wouldn;t that be enough or we want to have it part of cancellation event as well along with id?

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry_state.move:
```diff
@@ -0,0 +1,623 @@
+        let active_task_ids = vector[];
+        let ids = enumerable_map::get_map_list(&state.tasks);
+
+        vector::for_each(ids, |id| {
```
Since we are only reading the state, perhaps `vector::for_each_ref` can be used to avoid creation/copy of elements.

sjoshisupra replied:
@nizam-supraoracles

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry_state.move:
```diff
@@ -0,0 +1,623 @@
+        let automation_task_metadata = borrow_global<AutomationRegistryState>(@supra_framework);
+        if (enumerable_map::contains(&automation_task_metadata.tasks, id)) {
+            let value = enumerable_map::get_value(&automation_task_metadata.tasks, id);
+            value.state != PENDING
```
```suggestion
            value.state == ACTIVE
```
There is a `CANCELLED` state as well which would be missed by the existing logic

aregng replied:
As agreed, the task cancelled during epoch N is still active during that epoch, and only removed upon new epoch start.
so if node restarts in the middle of the epoch it should be able to get correct list of the tasks that are active during that epoch. that's why here we have `!= PENDING`

sjoshisupra replied:
Fair enough

Commit:
Addressed 2nd round of review comments

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry_state.move:
```diff
@@ -89,7 +92,7 @@ module supra_framework::automation_registry_state {
     #[event]
     /// Remove automation task registry event
-    struct RemoveAutomationTask has drop, store {
+    struct CanclledAutomationTask has drop, store {
```
`CancelledAutomationTask` (spelling)

Commit:
Fixed typo

Commit:
Addressed review comments
